### ISSUE TITLE
[Web] WebGPU issues tracking

### ISSUE THREAD

**Issue by fs-eire** [microsoft/onnxruntime#15796]
This issue is for tracking WebGPU related problems. WebGPU EP is available since ONNX Runtime Web v1.15.0 as experimental feature. We are working on improving stability, operator coverage and performance.

For a list of supported/WIP operators, comments or any operator specific issues: #15952

------------------

## Can not consume
Q: How to build?
A: Building ort-web with webgpu support from source: please refer to this [gist](https://gist.github.com/fs-eire/a55b2c7e10a6864b9602c279b8b75dce)

Q: [[Web] An error occurred during model execution: "TypeError: Cannot read properties of undefined (reading 'apply')".](https://github.com/microsoft/onnxruntime/issues/15719#top)
A: https://github.com/microsoft/onnxruntime/pull/15780 <--- this PR fixed it

Q: `no available backend found. ERR: ...`
A: Need to make sure webgpu is available in the current context. Upgrade to latest Chrome or Edge (v113), and served in a secured location ( https or localhost )

## Runtime failures

Q: `Non-zero status code returned while running Transpose node. ....`
A: #15819 <--- This PR should fix it

Q: crash in the transpose optimizer for various models (#15869: cannot load model https://huggingface.co/runwayml/stable-diffusion-v1-5/tree/onnx/vae_encoder)
A: issue being investigated - see the PR for detailed info

## Kernel coverage or running slow

Q: General investigation tips?
A: a few tools that can be used to taking deeper look at it: ( don't do them together, it will generate too many logs )
- `env.logLevel = 'verbose'; env.debug = true;` - This will let onnxruntime-web to output some logs helpful for analysing the execution. including telling which operators are running on webgpu and which are on CPU (fallback). to improve performance caused by fallback we need to improve the operator coverage. I can help to implement the missing ops.
- `env.webgpu.profilingMode = 'default';` - This will output quite a lot of logs into console for each webgpu shaders - by aggregating and analyzing those we can know which shader is slow. Need to launch chrome/edge with flag `--disable-dawn-features=disallow_unsafe_apis`.
- set `sessionOptions.enableProfiling = true` when creating inference session. This shows which operator running on GPU, which fallback to CPU.

Q: running slow on [image classification model](https://huggingface.co/Xenova/transformers.js/tree/main/quantized/google/vit-base-patch16-224/image-classification). ([logs](https://github.com/microsoft/onnxruntime/files/11354629/localhost-1682691588176.log))
A: `jsepCopyGpuToCpu` occurred 114 times, which indicating frequent CPU <--> GPU data transfer. Adding implementation of the missing operators may improve performance.

**Comment by xenova** [comment]
> For now, need chrome/edge canary launched with flag --enable-unsafe-webgpu, and served in a secured location ( https or localhost )

WebGPU is now supported in the latest version of the official Chrome build (no longer only Canary, and not locked behind the flag). That said, I do not know about support for other browsers.

**Comment by xenova** [comment]
> Q: Non-zero status code returned while running Transpose node. ....
> A: investigating. @xenova could you share the model and corresponding input?

[Here](https://huggingface.co/Xenova/transformers.js/tree/main/quantized/t5-small/seq2seq-lm-with-past) are the model files.

Here is some example input (for the encoder):
```js
let input = {
    attention_mask: new Tensor(
        'int64',
        new BigInt64Array([1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n]),
        [1, 12]
    ),
    input_ids: new Tensor(
        'int64',
        new BigInt64Array([13959n, 1566n, 12n, 2379n, 10n, 8774n, 6n, 149n, 33n, 25n, 58n, 1n]),
        [1, 12]
    )
}
```

**Comment by nagadomi** [comment]
>Q: Non-zero status code returned while running Transpose node. ....

I got the similar error.

```
23-05-04 08:10:11.420400 [W:onnxruntime:, session_state.cc:1171 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
i @ ort-wasm.js:49
23-05-04 08:10:11.427000 [E:onnxruntime:, sequential_executor.cc:514 ExecuteKernel] Non-zero status code returned while running Transpose node. Name:'/Transpose' Status Message: Failed to run JSEP kernel
Non-zero status code returned while running Transpose node. Name:'/Transpose' Status Message: Failed to run JSEP kernel
i @ ort-wasm.js:49
caught (in promise) Error: failed to call OrtRun(). error code = 1.
    at t.run (wasm-core-impl.ts:282:15)
    at async t.OnnxruntimeWebAssemblySessionHandler.run (session-handler.ts:81:11)
    at async d.run (inference-session-impl.js:91:15)
    at async Object.padding (script.js:498:19)
    at async Object.tiled_render (script.js:397:17)
    at async process (script.js:617:9)
    at async HTMLInputElement.<anonymous> (script.js:718:13)
```
The model(Object.padding) is very simple reflection padding.

PyTorch module
https://github.com/nagadomi/nunif/blob/6b605f8687b0ff5439f4c3776fd794f104cc2e15/nunif/models/onnx_helper_models.py#L12-L38
javascript caller
https://github.com/nagadomi/nunif/blob/6b605f8687b0ff5439f4c3776fd794f104cc2e15/waifu2x/unlimited_waifu2x/public_html/script.js#L483-L494
onnx file
[pad.zip](https://github.com/microsoft/onnxruntime/files/11391947/pad.zip)

**Comment by fs-eire** [comment]
The `Transpose` issue is figured out and a fix is in PR: #15819.

**Comment by fs-eire** [comment]
@DK013 The error message looks like the corresponding .wasm file is not served. please check from the devtool/network tab, if any 404 error occurs on *.wasm

**Comment by xenova** [comment]
Is there a list of supported WebGPU ops, as well as those planned to be implemented?

**Comment by fs-eire** [comment]
> Is there a list of supported WebGPU ops, as well as those planned to be implemented?

Let me update to the summary.

**Comment by fs-eire** [comment]
> > I think I can add a warning message if simd is off and webgpu is request.
> 
> @fs-eire Maybe along with the warning, you could ignore the `ort.env.wasm.simd` or reset it to `true`? Because warnings are often missed/ignored and the execution will still fail.

I will fail the initialization. see this PR: https://github.com/microsoft/onnxruntime/pull/15924

**Comment by xenova** [comment]
Are the versions released under the dev tag (e.g., https://www.npmjs.com/package/onnxruntime-web/v/1.16.0-dev.20230508-045c623415) built automatically from the main branch? This will mean I don't have to build the files myself for testing.

**Comment by fs-eire** [comment]
> @fs-eire Are the versions released under the dev tag (e.g., https://www.npmjs.com/package/onnxruntime-web/v/1.16.0-dev.20230508-045c623415) built automatically from the main branch? This will mean I don't have to build the files myself for testing.

They are, but it seems that the release pipeline is not doing perfectly. We are reworking the release pipeline for nightly builds recently. Before our work is done, you can use [this link](https://dev.azure.com/onnxruntime/onnxruntime/_build?definitionId=161&_a=summary&branchFilter=10261%2C10261%2C10261%2C10261%2C10261%2C10261%2C10261&statusFilter=succeeded%2Cwarning) to download latest artifacts from our public CI. hope this helps save your time.

**Comment by xenova** [comment]
Okay great! Can you perhaps just show where I can download the final builds? I'm not too familiar with the Azure DevOps UI.

**Comment by xenova** [comment]
So, I got the imports working, but I'm getting a lot of errors when running a simple text-classification [model](https://huggingface.co/Xenova/distilbert-base-uncased-finetuned-sst-2-english/blob/main/onnx/model.onnx):

input:
```js
{
    attention_mask: Tensor {
      type: 'int64',
      data: [1n, 1n, 1n],
      dims: [1,3],
    },
    input_ids: Tensor {
      type: 'int64',
      data: [101n, 3231n, 102n],
      dims: [1,3],
    },
}
```

![image](https://github.com/microsoft/onnxruntime/assets/26504141/9417dffc-6f00-47b2-956d-402b43bdb0a9)

**Comment by lxfater** [comment]
> ~Okay great! Can you perhaps just show where I can download the final builds? I'm not too familiar with the Azure DevOps UI.~ Nevermind, found it

How to download the release.

**Comment by xenova** [comment]
> How to download the release.

1. Go to [link](https://dev.azure.com/onnxruntime/onnxruntime/_build?definitionId=161&_a=summary&branchFilter=10261%2C10261%2C10261%2C10261%2C10261%2C10261%2C10261&statusFilter=succeeded%2Cwarning)
2. Click on build. e.g, https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1013308
3. Scroll down to "Stages"
4. Click "1 artifact" under "Build_web_Release"
![image](https://github.com/microsoft/onnxruntime/assets/26504141/f5b950af-7def-4769-95c3-7ad994b78916)

5. Click 3 dots on "NPM_packages"
![image](https://github.com/microsoft/onnxruntime/assets/26504141/17151414-f02c-42a0-8d46-b46d4a7fc090)

6. "Download artifacts"

**Comment by lxfater** [comment]
I’ve been trying to compile these files locally all day, but I keep getting an error: “Can’t create a session.” Even with the latest npm package, the issue persists. But the "onnxruntime-web": "^1.16.0-dev.20230508-045c623415" works well on wasn provider.

I’m using the Whisper model, and regardless, I sincerely appreciate your help. I hope you can integrate WebGPU to Transformers.js  soon  . Good night, my friend, from the other side of the Earth. 
@xenova

**Comment by vacing** [comment]
Met an error
```
gpu-data-manager.ts:168 Uncaught (in promise) Error: releasing data does not exist
    at s.release (gpu-data-manager.ts:168:13)
    at t.WebGpuBackend.free (backend-webgpu.ts:296:32)
    at Object.a [as Ra] (init.ts:110:34)
    at 879590 (ort-wasm-simd.jsep.js:20:122)
    at ee (ort-wasm-simd.jsep.js:29:128)
    at ort-wasm-simd.jsep.wasm:0xedb008
    at ort-wasm-simd.jsep.wasm:0x14b25
    at ort-wasm-simd.jsep.wasm:0x8c7ff9
    at ort-wasm-simd.jsep.wasm:0x4397ba
    at ort-wasm-simd.jsep.wasm:0x1e045c
```

**Comment by fs-eire** [comment]
> Met an error
> 
> ```
> gpu-data-manager.ts:168 Uncaught (in promise) Error: releasing data does not exist
>     at s.release (gpu-data-manager.ts:168:13)
>     at t.WebGpuBackend.free (backend-webgpu.ts:296:32)
>     at Object.a [as Ra] (init.ts:110:34)
>     at 879590 (ort-wasm-simd.jsep.js:20:122)
>     at ee (ort-wasm-simd.jsep.js:29:128)
>     at ort-wasm-simd.jsep.wasm:0xedb008
>     at ort-wasm-simd.jsep.wasm:0x14b25
>     at ort-wasm-simd.jsep.wasm:0x8c7ff9
>     at ort-wasm-simd.jsep.wasm:0x4397ba
>     at ort-wasm-simd.jsep.wasm:0x1e045c
> ```

@vacing could you share more details (share the model if possible)

**Comment by vacing** [comment]
> @vacing could you share more details (share the model if possible)
https://github.com/microsoft/onnxruntime/files/10138007/profiling.zip

this a demo version of my model, it can reproduce the error

**Comment by mrdomino** [comment]
I'm running into surprising/undesirable behavior passing `executionProviders: ['webgpu', 'wasm']` to `InferenceSession.create`: if I run on a mobile device, it fails with `WebGpuBackend: Failed to get GPU adapter.` This seems like it's seeing if webgpu works, assuming that it does prior to the failure, and trying to use it, rather than falling back to wasm.

Reading through the code a bit, it does seem like it's checking for `init()` success and trying the next backend after that, and I'm not seeing the error I'd expect if it were taking the "init failed" code path, so my guess is `init` is not initializing far enough to get to the error? I can't really tell because I don't understand the wasm proxy / message passing stuff well enough yet to know what's going on.

**Comment by mrdomino** [comment]
After further testing, it seems to be the case that merely importing `onnxruntime-web/webgpu` is sufficient to trigger the exception, even if `InferenceSession.create` is called with `executionProviders: ['wasm']`.

**Comment by gyagp** [comment]
@mrdomino, to make WebGPU work, you may just need to import ort.webgpu.min.js. 
I have some sample code at https://github.com/webatintel/ort-toolkit/blob/main/index.html#L109, and you may try a live demo at https://webatintel.github.io/ort-toolkit/?tasks=performance&ep=webgpu&modelName=mobilenetv2-12&modelUrl=hf&enableReadback=true

**Comment by mrdomino** [comment]
> @mrdomino, to make WebGPU work, you may just need to import ort.webgpu.min.js. I have some sample code at https://github.com/webatintel/ort-toolkit/blob/main/index.html#L109, and you may try a live demo at https://webatintel.github.io/ort-toolkit/?tasks=performance&ep=webgpu&modelName=mobilenetv2-12&modelUrl=hf&enableReadback=true

Thanks for the link! I'm now exploring further, and the behavior I'm seeing in Firefox (which I just installed) is different from the behavior I was seeing with Chrome on my Android phone. I'm going to focus on Firefox for now, as it's harder to debug on the phone.

First of all, it turns out that the import is not the issue after all — if I just pass `["wasm"]` as the executionProvider, it works whether I import onnxruntime-web or onnxruntime-web/webgpu (which is just an alias for ort.webgpu.min.js IIUC). 

If I just pass `["webgpu"]` on Firefox or Safari, I get `no available backend found. ERR: ` (with nothing after ERR.) This seems to be coming from [here](https://github.com/microsoft/onnxruntime/blob/main/js/common/lib/backend-impl.ts#L101), but it's surprising that there is no error message propagated up.

If I pass either `["webgpu", "wasm"]`, *or* `["wasm", "webgpu"]`, on Chrome, it works. (I cannot tell which execution provider it chose though.) But if I do so on either Firefox or Safari, I get: `i.Ea is not a function` with a reference into the minified ort.webgpu.min.js source.

I have thus far been working around this by manually testing `navigator.gpu.getAdapter()` and setting the backend based on that, but I'm confused as to why that is not happening automatically.

**Comment by mrdomino** [comment]
Okay, wow, this is getting complicated.

I just checked again with Chrome on Android, and even with my `navigator.gpu.requestAdapter()` workaround, I still get `WebGpuBackend: Failed to get GPU adapter.`, which appears to be coming from [here](https://github.com/microsoft/onnxruntime/blob/28ad3ff799e163bf5c854db03e87086a61e14256/js/web/lib/wasm/jsep/backend-webgpu.ts#L153). The only way to work around this is to make not just the executionProvider, but the *file import*, conditional on the result of `requestAdapter`. I'm not sure, but I bet your code has this problem too on Chrome on Android @gyagp: when I load the page there, the test results are never filled in with anything.

So to recap:

- Recent desktop Chrome: everything fine.
- Firefox/Safari on desktop: 'wasm' works, 'webgpu' does not work, and passing an array of backends causes things to break in a strange way, regardless of whether onnxruntime-web or onnxruntime-web/webgpu.
- Chrome on Android x onnxruntime-web: wasm works, webgpu presumably doesn't work (haven't tested since it seems pointless to)
- Chrome on Android x onnxruntime-web/webgpu: webgpu does not work, wasm does not work, and in both cases the error thrown is different from either of the Firefox/Safari errors.

**Comment by gyagp** [comment]
> @gyagp thanks for the pointers!
> 
> Your demo does run, but seems to throw some errors, so it's a little unclear what's being run on GPU vs CPU:
> 
> ```
> ort-wasm-simd.jsep.js:54 2023-10-27 10:51:47.156700 [W:onnxruntime:, session_state.cc:1162 VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
> j @ ort-wasm-simd.jsep.js:54
> ort-wasm-simd.jsep.js:54 2023-10-27 10:51:47.162400 [W:onnxruntime:, session_state.cc:1164 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
> ```
> 
> ![image](https://user-images.githubusercontent.com/493968/278722402-7be249db-ceb5-4f68-ad5a-b10eef790991.png)
> 
> I think the suggestion to "rerun with verbose output on a non-minimal build" to show node assignments requires you to re-build, right? Is that something you're able to do?
> 
> Thanks again!

@gabrielgrant, these errors could be ignored now. If we use webgpu as ep, if some ops are not supported by WebGPU, they would automatically fall back to wasm. My script has a ortProfiling task (Whole url is https://webatintel.github.io/ort-toolkit/?tasks=ortProfiling&ep=webgpu&modelName=mobilenetv2-12&modelUrl=hf&enableReadback=true), and it will show you where each op is running on (JsExecutionProvider means WebGPU, while CPUExecutionProvider means wasm).

**Comment by gyagp** [comment]
> Okay, wow, this is getting complicated.
> 
> I just checked again with Chrome on Android, and even with my `navigator.gpu.requestAdapter()` workaround, I still get `WebGpuBackend: Failed to get GPU adapter.`, which appears to be coming from [here](https://github.com/microsoft/onnxruntime/blob/28ad3ff799e163bf5c854db03e87086a61e14256/js/web/lib/wasm/jsep/backend-webgpu.ts#L153). The only way to work around this is to make not just the executionProvider, but the _file import_, conditional on the result of `requestAdapter`. I'm not sure, but I bet your code has this problem too on Chrome on Android @gyagp: when I load the page there, the test results are never filled in with anything.
> 
> So to recap:
> 
> * Recent desktop Chrome: everything fine.
> * Firefox/Safari on desktop: 'wasm' works, 'webgpu' does not work, and passing an array of backends causes things to break in a strange way, regardless of whether onnxruntime-web or onnxruntime-web/webgpu.
> * Chrome on Android x onnxruntime-web: wasm works, webgpu presumably doesn't work (haven't tested since it seems pointless to)
> * Chrome on Android x onnxruntime-web/webgpu: webgpu does not work, wasm does not work, and in both cases the error thrown is different from either of the Firefox/Safari errors.

@mrdomino I'm not sure about the exact status of Firefox and Safari, but for Chrome, WebGPU was only officially supported on Windows, macOS and ChromeOS (since M113). Android support is still behind the flag "--enable-unsafe-webgpu". Fortunately the its status is very good now, as Google just sent out "intent to ship" in Chrome (https://chromestatus.com/feature/5119617865613312), and planned to ship it in M121 (Jan 23 is the release date). So before then, you still need to pass switch "--enable-unsafe-webgpu" to enable WebGPU on Android with latest Chrome (better to experiment with Chrome Canary).

**Comment by gyagp** [comment]
@mrdomino BTW, if you have interest to follow up the WebGPU status on Safari, here has some clue: https://github.com/gpuweb/gpuweb/issues/4238

**Comment by mrdomino** [comment]
I'm actually not that worried about webgpu on Safari — using pure wasm as a fallback is acceptable for my use case, and actually works well enough on Firefox.

The things I'm concerned about are basically just:

1. Having to conditionally import one of two different headers depending on navigator.gpu is painful, particularly with TypeScript — it'd be nice if one header worked in both situations.
2. The inconsistent fallback / failure behavior is surprising, and seems like it's unintentional.

**Comment by gyagp** [comment]
> I'm actually not that worried about webgpu on Safari — using pure wasm as a fallback is acceptable for my use case, and actually works well enough on Firefox.
> 
> The things I'm concerned about are basically just:
> 
> 1. Having to conditionally import one of two different headers depending on navigator.gpu is painful, particularly with TypeScript — it'd be nice if one header worked in both situations.

You only need to import ort.webgpu.min.js. Then if WebGPU is supported, use webgpu as ep; otherwise, change ep to wasm. Pseudo code:
if (webgpuSupported) {
  ep = 'webgpu';
} else {
  ep = 'wasm';
}
const option = {
  executionProviders: [
    {
      name: ep,
    }
  ]
};
const session = await createSession(option);

> 2. The inconsistent fallback / failure behavior is surprising, and seems like it's unintentional.
If WebGPU is supported, the fallback to wasm is either a limitation (some ops, including dataType variants, are not implemented by WebGPU) or an optimization (ORT has heuristic to use wasm over WebGPU for better performance). We will continue to improve the framework, including the profiling mechanism, so that it's easier to differentiate these two. You're always welcome to report perf issue when in doubt.

**Comment by mrdomino** [comment]
> You only need to import ort.webgpu.min.js. Then if WebGPU is supported, use webgpu as ep; otherwise, change ep to wasm.

No, that is not the case in Chrome on Android. That works on Chrome on Desktop, and on Firefox and Safari on desktop, but not on Android. On Android, no matter what ep is, importing ort.webgpu.min.js causes a crash.

And by point 2, I was referring to the strange error message thrown on Firefox/Safari (`i.Ea is not a function`) when I pass an array of backend hints.

**Comment by mrdomino** [comment]
The specific code I am using to decide which backend to use is:

```javascript
const backend = await (async () => {
  if (!navigator.gpu) return 'wasm'
  const adapter = await navigator.gpu.requestAdapter()
  if (!adapter) return 'wasm'
  return 'webgpu'
})()
```

That backend is then passed in to InferenceSession.create (as `{ executionProviders: [backend], ...}`). Depending on which javascript file is imported, that either crashes with "`WebGpuBackend: Failed to get GPU adapter.`" or works.

I will try to get a code sandbox up with a minimal example.

**Comment by mrdomino** [comment]
Here: https://ort-test.vercel.app/

The only difference between the WASM and WebGPU pages is which file is imported. Both are using "wasm" as the ep. On Chrome Android, WASM says "Everything worked" and WebGPU says "Failed during InferenceSession.create" with the WebGpuBackend error message.

https://bitbucket.org/mrdomino/ort-test/src/main/app/ort/page.tsx
https://bitbucket.org/mrdomino/ort-test/src/main/app/ort-webgpu/page.tsx

**Comment by mrdomino** [comment]
Curiously, Chrome on Android exposes a `navigator.gpu`. It just doesn't produce an adapter if you request one. Firefox and Safari both do not have a `navigator.gpu` at all. (I added text to the test app's main page to distinguish this.)

Is it possible that somewhere in the code there is a simple check for the presence of navigator.gpu to decide to use WebGPU?

**Comment by mrdomino** [comment]
It looks like `registerBackend` has such a check:

https://github.com/microsoft/onnxruntime/blob/24f9c1afe3a1972cfacd9b4bcd41f2227d369bf0/js/web/lib/index.ts#L18-L23

So there is a difference between the good cases of Firefox/Safari and the bad case of Android Chrome: the former do not have the backend registered, while the latter does.

Still, the logic in resolveBackend really looks like it should be handling the error, and from the observed behavior, it is not. So a different code path must be getting taken that is trying to initialize a WebGpuBackend.

Ah, and indeed, here we go:

https://github.com/microsoft/onnxruntime/blob/24f9c1afe3a1972cfacd9b4bcd41f2227d369bf0/js/web/lib/wasm/jsep/init.ts#L133-L140

I think I can probably submit a PR to fix that.

**Comment by gyagp** [comment]
Thanks for the PR, @fs-eire and @guschmue, any comments on this?
@mrdomino It's also easy for you to work around this at your side like below:
  async function getEp() {
    if (!navigator.gpu) {
      return 'wasm';
    }
    const adapter = await navigator.gpu.requestAdapter();
    if (!adapter) {
      **delete Navigator.prototype.gpu;**
      return 'wasm';
    }
    return 'webgpu';
  }

**Comment by mrdomino** [comment]
Hahaha, I like it.

One thought I had is it probably makes sense to also check the adapter at the register backend call site, and in that case, maybe it makes sense for hasGpu to be in the environment?

**Comment by mrdomino** [comment]
FYI, using onnxruntime-web 1.17.0, `executionProviders: ['webgpu', 'wasm']` still does not work on Firefox — it dies with the same `[minified-name]` is not a function` as before.

**Comment by alba-saco** [comment]
Hi there,

I'm getting this error with webgpu executionProvider:
`audioProcessor.js:2 Error during inference: Error: no available backend found. ERR: [webgpu] RuntimeError: null function or function signature mismatch`

When I remove `ort.env.wasm.simd = false` I get the following error:
`Error during inference: TypeError: Cannot read properties of undefined (reading 'apply')`
To confirm, I am importing from `onnxruntime-web/webgpu` and am running on Chrome via https.
When I run my code with executionProviders=['wasm'] everything executes perfectly.

I wasn't sure if I should create a new issue or just put a comment here

**Comment by fs-eire** [comment]
> Hi there,
> 
> I'm getting this error when I set executionProviders=['webgpu'] (I am running on Chrome via https): `audioProcessor.js:2 Error during inference: Error: no available backend found. ERR: [webgpu] RuntimeError: null function or function signature mismatch`
> 
> When I remove `ort.env.wasm.simd = false` I get the following error: `Error during inference: TypeError: Cannot read properties of undefined (reading 'apply')` I am aware of [this issue](https://github.com/microsoft/onnxruntime/issues/15719#top) and to confirm, I am importing from `onnxruntime-web/webgpu`
> 
> When I run my code with executionProviders=['wasm'] everything executes perfectly.
> 
> I wasn't sure if I should create a new issue or just put a comment here

This looks like you may import from `onnxruntime-web` not `onnxruntime-web/webgpu` , see https://github.com/microsoft/onnxruntime-inference-examples/tree/main/js/importing_onnxruntime-web#conditional-importing

**Comment by fs-eire** [comment]
> Thank you for your prompt response. I get `Error during inference: Error: no available backend found. ERR: ` when importing from `onnxruntime-web`. I am running on the latest version of chrome served with https and have confirmed that webGPU is available in my environment

Yes, as I explained, if you want to use WebGPU, you need to import `onnxruntime-web/webgpu`. Otherwise it does not work.

**Comment by guschmue** [comment]
closing this one, gotten a bit stale.
For webgpu issues lets create new issues for each problem.
